packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModule.kt

```diff
@@ -0,0 +1,326 @@
+    }
+    return database
+  }
+
+  private class SQLitePluginResult(
+    val rows: Array<Array<Any?>?>,
+    val columns: Array<String?>,
+    val rowsAffected: Int,
+    val insertId: Long,
+    val error: Throwable?
+  )
+
+  private class ReadOnlyException : Exception("could not prepare statement (23 not authorized)")
+  companion object {
+    private const val DEBUG_MODE = false
+    private val TAG = SQLiteModule::class.java.simpleName
+    private val EMPTY_ROWS = arrayOf<Array<Any?>?>()
+    private val EMPTY_COLUMNS = arrayOf<String?>()
+    private val EMPTY_RESULT = SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, 0, 0, null)
+    private val DATABASES: MutableMap<String, SQLiteDatabase?> = HashMap()
+
+    @Throws(IOException::class)
+    private fun ensureDirExists(dir: File): File {
+      if (!dir.isDirectory) {
+        if (dir.isFile) {
+          throw IOException("Path '$dir' points to a file, but must point to a directory.")
+        }
+        if (!dir.mkdirs()) {
+          var additionalErrorMessage = ""
+          if (dir.exists()) {
+            additionalErrorMessage = "Path already points to a non-normal file."
+          }
+          if (dir.parentFile == null) {
+            additionalErrorMessage = "Parent directory is null."
+          }
+          throw IOException("Couldn't create directory '$dir'. $additionalErrorMessage")
+        }
+      }
+      return dir
+    }
+
+    private fun pluginResultsToPrimitiveData(results: Array<SQLitePluginResult?>): List<Any> {
+      val list: MutableList<Any> = ArrayList()
+      for (i in results.indices) {
+        val result = results[i]
+        val arr = convertPluginResultToArray(result)
+        list.add(arr)
+      }
+      return list
+    }
+
+    private fun convertPluginResultToArray(result: SQLitePluginResult?): List<Any?> {
+      val data: MutableList<Any?> = ArrayList()
+      data.add(result!!.error?.message)
+      data.add(result.insertId.toInt())
+      data.add(result.rowsAffected)
+
+      // column names
+      val columnNames: MutableList<String?> = ArrayList()
+      for (i in result.columns.indices) {
+        columnNames.add(result.columns[i])
+      }
+      data.add(columnNames)
+
+      // rows
+      val rows: MutableList<Any> = ArrayList()
+      for (i in result.rows.indices) {
+        val values = result.rows[i] ?: continue
+
+        // row content
+        val rowContent: MutableList<Any?> = ArrayList()
+        for (j in values.indices) {
+          val value = values[j]
+          if (value == null) {
+            rowContent.add(null)
+          } else if (value is String) {
+            rowContent.add(value as String?)
+          } else if (value is Boolean) {
+            rowContent.add(value as Boolean?)
+          } else {
+            val v = value as Number
+            rowContent.add(v.toDouble())
+          }
+        }
+        rows.add(rowContent)
+      }
+      data.add(rows)
+      return data
+    }
```

Comment:
1. result shouldn't be nullable 
2. Can it be rewritten using a functional paradigm? Using map et.